### PR TITLE
1271 access-policy use jest framwork only

### PR DESCRIPTION
**Pull request by billli0**
### Issue #:
[1271- Use a single test framework : **access-policy module only**](https://github.com/beyondessential/tupaia-backlog/issues/1271)

This branch is a draft PR for tracking the changes.

### CODE REVIEW

**Comment by billli0** [comment]
I'll adjust a few points to stick with the testing standard.

File: packages/access-policy/scripts/JestTransformer.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/**
+ * https://babeljs.io/docs/en/config-files#jest
+ *
+ * Setup jest in Monorepo. This is to create a custom jest transformer file that wraps babel-jest.
+ *
+ * @type {Transformer}
+ */
+module.exports = require('babel-jest').createTransformer({
+  rootMode: 'upward',
+});
```
Comment (kael89) [inline packages/access-policy/scripts/JestTransformer.js:10]: Let's try defining this under the monorepo root, and importing it where required! No need to do this in this PR though, let's submit another PR for that

File: packages/access-policy/src/__tests__/AccessPolicy.test.js [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        ['entity (i)', ['KI'], ['Public', 'Admin']],
+        ['entity (ii)', ['DL', 'KI'], ['Public', 'Admin']],
+        [
+          'entity (iii)',
+          ['SB', 'KI'],
+          ['Public', 'Admin', 'Royal Australasian College of Surgeons'],
+        ],
+      ];
+
+      it.each(testData)('%s', (_, arr, expected) => {
```
Comment (kael89) [inline packages/access-policy/src/__tests__/AccessPolicy.test.js:121]:
```suggestion
      it.each(testData)('%s', (_, entities, expected) => {
```

And in the line below

File: packages/access-policy/src/__tests__/AccessPolicy.test.js [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+      ];
+
+      it.each(testData)('%s', (_, [entities, permissionGroup], expected) => {
+        expect(accessPolicy.allowsSome(entities, permissionGroup)).toBe(expected);
+      });
+    });
+  });
+
+  describe('getPermissionGroups', () => {
+    it('should return all permission groups when no entities are specified', () => {
```
Comment (kael89) [inline packages/access-policy/src/__tests__/AccessPolicy.test.js:104]: This test case can be consolidated with the parameterized tests below:

```js
[
  'should return all permission groups when no entities are specified',
  undefined,
  ['Public', 'Admin', 'Royal Australasian College of Surgeons'],
],
```

File: packages/access-policy/src/__tests__/AccessPolicy.test.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+      expect(accessPolicy.getPermissionGroups()).toStrictEqual(
+        expect.arrayContaining(['Public', 'Admin', 'Royal Australasian College of Surgeons']),
+      );
+    });
+
+    describe('should return just the permission groups that relate to the provided entities', () => {
```
Comment (kael89) [inline packages/access-policy/src/__tests__/AccessPolicy.test.js:110]: And then we can delete this describe block (using its description in the parameterized test cases below)

File: packages/access-policy/src/__tests__/AccessPolicy.test.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  describe('allowsSome', () => {
+    describe('no permission group', () => {
+      const testData = [
+        [
+          'should allow access to an entity that has at least one permission group, when no permission group is specified',
+          ['DL'],
+          true,
+        ],
+        ['should not allow access to an entity that has no permission groups', ['VU'], false],
+        ['should not allow access with no entity', [], false],
```
Comment (kael89) [inline packages/access-policy/src/__tests__/AccessPolicy.test.js:53]: We also need one test case for `undefined`

```suggestion
        ['should not allow access with no entity', [], false],
        ['should not allow access with no entity', undefined, false],
```

This corresponds to the following test case in the existing code:

```js
expect(accessPolicy.allowsSome()).to.equal(false);
```

**Review by kael89 — CHANGES_REQUESTED** [review]
Great work! A few details here and there, but looking good overall!

File: packages/access-policy/src/__tests__/AccessPolicy.test.js [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        'should allow access to a permission group when a matching entity is specified',
+        ['DL', 'Public'],
+        true,
+      ],
+      [
+        'should not allow access to a permission group that exists for a different entity than the one specified',
+        ['DL', 'Admin'],
+        false,
+      ],
+      ['should not allow access with no entity', ['', ''], false],
```
[the remark this answers is not in the document]
Comment (billli0) [reply packages/access-policy/src/__tests__/AccessPolicy.test.js:35]: I put that in` line 82`. Hope that is correct.

[the remark this answers is not in the document]
Comment (billli0) [reply packages/access-policy/src/__tests__/AccessPolicy.test.js:35]: > One case is missing here. Existing code:
> 
> ```js
>     it('should allow access to a permission group that includes spaces', () => {
>       expect(accessPolicy.allows('SB', 'Royal Australasian College of Surgeons')).to.equal(true);
>     });
> ```

I put that in line 82. Hope that is correct.

[the remark this answers is not in the document]
Comment (kael89) [reply packages/access-policy/src/__tests__/AccessPolicy.test.js:35]: This uses `accessPolicy.allows`, while in line 82 it uses `accessPolicy.allowsSome`. Both cases are required

[the remark this answers is not in the document]
Comment (billli0) [reply packages/access-policy/src/__tests__/AccessPolicy.test.js:35]: > This uses `accessPolicy.allows`, while in line 82 it uses `accessPolicy.allowsSome`. Both cases are required

You are right. My bad. I'll add that.

[the remark this answers is not in the document]
Comment (kael89) [reply packages/access-policy/src/__tests__/AccessPolicy.test.js:35]: No worries! 😃

File: packages/access-policy/src/__tests__/AccessPolicy.test.js [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+          ],
+          [
+            ['SB', 'KI'],
+            ['Public', 'Admin', 'Royal Australasian College of Surgeons'],
+          ],
+        ],
+      ],
+    ];
+
+    it.each(testDatas)('%s', (_, testData) => {
```
Comment (kael89) [inline packages/access-policy/src/__tests__/AccessPolicy.test.js:134]:
```suggestion
    it.each(testData)('%s', (_, testCaseData) => {
```

(You'll have to rename the corresponding vars in this block accordingly)

**Review by kael89 — APPROVED** [review]
Nice one! Just a detail, but pre-approving

**Comment by billli0** [comment]
> Nice one! Just a detail, but pre-approving

Awesome! I will merge it after test pass.

**Comment by kael89** [comment]
@billli0 good call, we can merge this directly because only tests were affected (no production code)

**Pull request merged**